### DISCUSSION TITLE
Best way to “upgrade” files for existing albums?

### DISCUSSION THREAD

**Discussion by justinmayer** [beetbox/beets#4755] [Q&A] [answered]
I occasionally find myself in situations like the following:

* I have an existing album in my library in MP3 format but want to replace it with new FLAC-format versions of the exact same tracks.
* I have an existing album in my library in FLAC format but want to replace it with a FLAC album that contains a bonus track not present on the existing album in my library.
* I have an existing “deluxe version” of an album in my library in FLAC format but want to replace it with a non-deluxe version of that album that contains fewer tracks than the existing album in my library.
* I have an existing album in my library in FLAC format but want to replace it with a “better” FLAC version that contains the exact same set of tracks.

In each case, none of the files in the existing album will remain behind, since the entire album will be replaced with the newer version.

Do y’all have suggestions for recommended best practices for these use cases? What exact Beets invocation(s) should I use to address these use cases?

**Comment by arsaboo** [comment]
There is a PR https://github.com/beetbox/beets/pull/4471 in the works to precisely handle this issue.

**Reply by justinmayer** [reply]
Yes, I am aware, and I look forward to seeing what it might enable if and when that PR lands, but I am curious what folks are doing with the current version of Beets to address these use cases.

**Reply by arsaboo** [reply]
Unfortunately, there is nothing else. I normally copy and reimport all the files again.

**Reply by justinmayer** [reply]
Do you delete the existing files in the library first? What do you mean when say you "copy and reimport all the files again"? What is the precise process that you use?

**Reply by arsaboo** [reply]
I copy the old files again with the new files (remove any duplicate/lower quality files) in a folder and re-import them.

**Reply by justinmayer** [reply]
It turns out that a standard `beet import […]` will handle these use cases. I appreciate you taking the time to reply, Alok. Much appreciated!

**Comment by justinmayer** [comment, marked as answer]
After some experimentation, I discovered that running `beet import /path/to/new/version/of/same/album` will yield:

```
This album is already in the library!
Old: 11 items, MP3, 249kbps, 43:58, 78.2 MiB
New: 11 items, FLAC, 890kbps, 44.1kHz/16 bit, 43:58, 275.1 MiB
[S]kip new, Keep all, Remove old, Merge all?
```

Typing `R` and then the `Return` key will remove the old version of the album and replace it with the new version. Much simpler than I realized! 😁

**Reply by arsaboo** [reply]
Yes, the problem with that is that even if you add one new file, all the old files will be removed and you will only have 1 file in the reimported album.

**Reply by justinmayer** [reply]
As noted in my use cases above, up until now I have only ever needed to replace entire versions of albums, replete with all tracks, so I have never needed to retain any tracks from the old versions of the albums that I am replacing. That said, if I ever do have that need, I imagine the "Merge all" option might address the use case you describe.

**Reply by JOJ0** [reply]
Glad you found a workflow that works for you :-) The merge all option does not exactly as you would have imagined. It tries to "build" an album out of all the files you already have and add those tracks you are adding newly. Well, sounds complicated. I'll try with an example:

- You have a 10 track album in your library
- You only have 7 of the tracks, your album is incomplete
- You import the 3 missing tracks (select _exactly_ the album you have tagged the (7/10) album before)
- Say merge at that prompt. The importer now takes all the 10 tracks and searches with this additional information again to find a good match. Usually you will find exactly that album you had tagged the 7/10 album before.
- Done

What the merge as well as the replace optoion can't do is what @arsaboo describes. You have 10/10 tracks in the album and you want to replace 2 files for example. If you'd select merge, weird things might happen, the importer wants to find an album consisting of 12 tracks. It does not get that 2 files are actually duplicates. Also weird to explain. Maybe you want to give it a try to understand.

I'm brainstorming at the moment if in reality the merge option should be improved or if a new option like upgrade is the way to go.....just thinking out loud....might be a bad idea.....

**Comment by arsaboo** [comment]
@JOJ0 actually, in the scenario you mentioned (we have 10/10 tracks and we want to merge two new files), Beets just creates another file with .1 added. 

So merge behaves as it should. I feel `upgrade` is the way to go to upgrade the files. To keep things simple, we can provide a config option to provide the default action (say upgrade) when duplicates are encountered.
